### PR TITLE
Upgrade React packages to v18

### PR DESCRIPTION
**Pull request by jsnajdr**
Subset of #32765 that only upgrades the React packages, and does no other changes. I'd like to see which e2e tests etc are failing on the latest `trunk`.

### CODE REVIEW

File: packages/block-directory/package.json [hunk 1]
```diff
@@ -47,8 +47,8 @@
 		"change-case": "^4.1.2"
 	},
 	"peerDependencies": {
-		"react": "^17.0.0",
-		"react-dom": "^17.0.0"
+		"react": "^18.0.0",
```
Comment (noahtallen) [inline packages/block-directory/package.json:50]: In order for the rest of the ecosystem to update smoothly and at their own pace, should we support both 17 and 18 as peer deps in most packages? (As long as our code is still compatible with 17, that is)

Comment (jsnajdr) [reply packages/block-directory/package.json:50]: I'm not sure if our code will be compatible, that's the problem.

For example, I suspect that we'll need to update the `useSelect` hook to use `useSyncExternalStore` from React 18, otherwise it will not work 100% reliably with v18. Then the `@wordpress/data` library has a non-negotiable dependency on React 18, and that transitively affects all other libraries...

Let's keep the v17 compat requirement in mind, and see what we can do. Libraries like `react-redux` use a compat shim for `useSyncExternalStore`, and that allows them to support older versions back to React 16.

**Comment by jsnajdr** [comment]
By far the biggest issue with the React 18 migration are unit tests firing warnings from `react-test-renderer` (and in turn from `@testing-library/react`):
```
An update to SomeComponent inside a test was not wrapped in act(...)
```
These warnings were quite common in React 17, too, but now React 18 seems to be much more sensitive about them, probably because of the concurrent mode, and tests often fail.

This `act()` issue is a major mess because it's quite hard to understand deeply what `act()` does, what kind of component behavior triggers the warnings and what's the correct way to avoid them.

In most affected libraries, someone filed a GitHub issue and most of the time the issue is closed after some time because most participants are confused and nobody is able to pinpoint the exact cause:
- https://github.com/callstack/react-native-testing-library/issues/379
- https://github.com/react-navigation/react-navigation/issues/9701
- https://github.com/testing-library/react-testing-library/issues/1051

The `@floating-ui/react-dom` library [recommends a certain workaround in its docs](https://floating-ui.com/docs/react-dom#testing) which kind of works, but it's a hack and is not entirely correct.

Tests are failing on `act()` when there is some async (i.e., executed in a `Promise.then` handler) update scheduled by the tested React UI.

Most typical case is fetching some data with `@wordpress/data`. This is Gutenberg code and we usually handle that by mocking the `@wordpress/data` calls, intercepting the promise that does the fetch and awaiting it inside `act()`:
```js
const fetchPromise = new Promise( ( resolve ) => {
  apiFetch.mockImplementation( async ( { path } ) => {
    resolve();
    return response;
  } );
} );

await act( () => fetchPromise );
```
This works but it requires access to the library internals (or mocking the internals). And it's not reliable: it relies on the fact that the `act` call finishes in the same or later event loop tick as the React update triggered by the finished fetch. Just one extra `setTimeout( 0 )` will break it.

There are libraries where the async update is invisible: the `useFloating` hook from `@floating-ui/react-dom` will do something like:
```js
function useFloating() {
  const [ data, setData ] = useState( { x: null, y: null } );
  useEffect( () => {
    computeDomPosition().then( ( data ) => setData( data ) );
  }, [] );
  return data;
}
```
I.e., it will calculate the DOM position of the rendered position and updates state in a `.then` handler. A component that uses `useFloating` will first render with default position (`{ x: null, y: null }`) and then, after a tick, will rerender with updated position.

But then a synchronous test that renders this component will miss the update:
```js
render( <FloatingComponent title="floating title" /> );
expect( screen.getByTitle( 'floating title' ) ).toBeInTheDocument();
```
It will work (probably) because the assertion doesn't depend on positioning, but the state update will run a bit later, possibly during another test, and will trigger the "not wrapped in act()" warning.

A common solution (recommended even by the library docs) is to wait for the extra tick:
```js
render( <FloatingComponent title="floating title" /> );
await act( async () => {} );
expect( screen.getByTitle( 'floating title' ) ).toBeInTheDocument();
```
But this is next to impossible to do reliably: every time the tested UI does call `useFloating` that changes position, I'm supposed to write a precisely placed `await act()`? I was able to do this for several tests in this PR, but only by stepping through the test suites in a debugger and tracking down places that trigger these updates.

Another library that performs the `useEffect` + `Promise` combo is `react-navigation`, a navigation/routing library for React Native, whose `NavigationContainer` component does this to get the initial state of the route. Some of our React Native unit tests do the same spurious await, like this one:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/test/edit.native.js#L85-L89

**The Right Solution**
The right solution, consistent with the Testing Library principles, is to realize that every async state update, be it a fetch, floating positioning or navigation init, should have some effect on the rendered UI. Placeholder disappears, element has an updated position, `<NavigationContainer>` renders its children instead of a placeholder.

And our tests need to _wait_ for that update to happen. For example, for `useFloating` we'll probably need to write a Jest helper that does something like:
```js
await waitFor( () => hasFinalPosition( screen.getByRole( 'popover' ) ) );
```
where `hasFinalPosition` will inspect the relevant DOM element style and check the positioning.

**Comment by jsnajdr** [comment]
> For example, for `useFloating` we'll probably need to write a Jest helper that does something like:

I just implemented this in #45587 and it really works :tada:

**Comment by tyxla** [comment]
There's a failing test for the `MediaReplaceFlow` component in this branch.

I've created #45736 to address it separately.

File: packages/block-editor/src/components/block-vertical-alignment-control/test/index.js [hunk 2]
```diff
@@ -3,6 +3,7 @@
  */
 import { render, screen } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
+import { act } from 'react-test-renderer';
```
Comment (tyxla) [inline packages/block-editor/src/components/block-vertical-alignment-control/test/index.js]: @jsnajdr perhaps we could use the re-exported `act` from `@testing-library/react`?

File: packages/block-editor/src/components/link-control/test/index.js [hunk 3]
```diff
@@ -245,7 +244,8 @@ describe( 'Basic rendering', () => {
 				name: 'Edit',
 			} );
 
-			await user.click( editButton );
+			fireEvent.click( editButton );
```
Comment (tyxla) [inline packages/block-editor/src/components/link-control/test/index.js]: I'm curious about those instances where we actually go back and refactor from `user-library` to `fireEvent`. Could you elaborate on the rationale for that?

Comment (jsnajdr) [reply packages/block-editor/src/components/link-control/test/index.js]: I'm not sure, but I think it was something like this: `user.click` is an async loop like:
```js
for ( const action of pointerActions ) {
  await wait() // promisified setTimeout( 0 )
  act( () => pointerAction( action ) );
}
```
I.e. there is an async loop and the `act()` calls are synchronous. So, the update from `computePosition().then()` falls in between the `pointerAction` calls and is not wrapped in `act()`.

But it's possible it's all bogus and `user.click` is fine. We'll need to revisit this. It's been a few weeks since I originally did these particular changes and today I know much more about how async `act()` works.

File: packages/block-editor/src/components/block-vertical-alignment-control/test/index.js (hunk 2 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/block-vertical-alignment-control/test/index.js]: Oh yes, this was VSCode eagerly adding a random import statement as soon as I typed "act" 🙂

**Comment by tyxla** [comment]
Addressing the block alignment tests separately in #45937.

**Comment by tyxla** [comment]
@jsnajdr I've downgraded `react-native` to `0.69.4` since that's the version we're aiming for AFAIK - see #43485. 

cc @geriux

**Comment by jsnajdr** [comment]
> Update native snapshots

@tyxla I was curious where these `nativeID="animatedComponent"` props are coming from and why they disappeared. It was the `Animated.View` component, part of the `react-native` package, and the `react-native` upgrade changed the `Animated.View` implementation to no longer use the `nativeID` prop: https://github.com/facebook/react-native/commit/4fdbc44ab5945399338e4ed94ea5611098bd2142

**Comment by jsnajdr** [comment]
> I've downgraded react-native to 0.69.4 since that's the version we're aiming for

Thank you, I just squashed that commit into the previous "upgrade React Native" one. So that we have a series of commits where every change is done only once. Similarly, I squashed the commit that removes the unwanted `act from 'react-test-renderer'` imports into the one that introduced them.

It's important for keeping the PR branch in a comprehensible state, and for preventing rebases from becoming a nightmare.

**Comment by jsnajdr** [comment]
Current state of this PR is that all unit tests, both web and mobile, are green. Only mobile e2e tests are failing:
- the Android version fails the `should be able to create a post with all blocks and scroll to the last one` test
- the iOS version fails to build the editor and doesn't even get to actual e2e testing

But @fluiddot promises us that these mobile tests will get fixed after the separate React Native upgrade in #43485 gets merged, so we don't need to worry about them.

There are however some things that are risky and that could lead to the editor being semi-broken after the React 18 upgrade.

The `useNestedSettingsUpdate` hook is used by each `InnerBlock` component instance (i.e., every block that has inner blocks) to dispatch an `updateBlockListSettings` action when the settings change. To avoid doing this for each block separately, the hook is optimized so that it adds the update to a queue, and then in the next microtask tick does all the updates in a batch. Unfortunately, this leads to the already familiar `useEffect( () => { promise.then( () => setState() ) } )` pattern that leads to a state update not wrapped in `act()`. I solved this for now by deoptimizing, i.e., dispatching `updateBlockListSettings` immediately.

I suspect this breaks the editor a little bit. I've seem an intermittent e2e failure of the "Quote can be unwrapped with Backspace" e2e test. That's suspicious because Quote block has inner blocks, and there is this code comment by @youknowriad on the `useInnerBlockTemplateSync` hook saying that it's closely related to `useNestedSettingsUpdate` and that one way it could fail is ... inserting a Quote block 🙂  https://github.com/WordPress/gutenberg/blob/ec2bdc59c4c4cdf8d86cdd432ee30eaca6f14eed/packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js#L56-L59

Then there are mobile unit tests that still trigger the "update not wrapped in `act()`" warnings. But because unexpected console warnings don't lead to a test failure in the mobile suite, and because the warning don't affect the results of the tests, the unit test suite comes back green. But it's a dangerous situation and the test can intermittently fail. We should fix them as soon as possible after merging the migration.

I've seen several instances of this warning. One of them is a `isScreenReaderEnabled` check like this:
```js
useEffect( () => {
  AccessibilityInfo.isScreenReaderEnabled().then( ( screenReaderEnabled ) => {
    setIsScreenReaderEnabled( screenReaderEnabled );
  } );
}, [] );
```
Again, setting state in effect + promise. It's interesting that the `act()` warning happens even though the `screenReaderEnabled` state update is guaranteed to be from `false` (initial state value) to `false` (mocked return value of `isScreenReaderEnabled()`), .e.g., always a noop.

Another warning happens when dispatching the `__experimentalTearDownEditor` action when unmounting the editor. The state update triggers all the `useSelect` callbacks, including ones on already unmounted components (!) and the `select` calls in these callback can trigger selector resolution.

Fixing this might force us to rewrite the `useSelect` hook to `useSyncExternalStore`. In the current implementation I had to disable the "avoid calling nested listener after unmounted" unit test, which is impossible to fix in React 18 without the `useSyncExternalStore` hook.

**Comment by youknowriad** [comment]
> Fixing this might force us to rewrite the useSelect hook to useSyncExternalStore. In the current implementation I had to disable the "avoid calling nested listener after unmounted" unit test, which is impossible to fix in React 18 without the useSyncExternalStore hook.

Did you experiment with this yet? I recall the last time I tried, the performance was very bad 😬 so I kind of nervous about this.

**Comment by jsnajdr** [comment]
> Did you experiment with this yet? I recall the last time I tried, the performance was very bad

Not yet, I prioritized fixing the non-negotiable things first. But I'll probably get to work on it soon.

**Comment by jsnajdr** [comment]
@geriux So I cherry-picked all the commits from #43485 and it's been a great progress! The iOS e2e test suite is green, only the Android tests are failing. On this:
```
FAIL __device-tests__/gutenberg-editor-initial-html.test.js (134.58 s)
  Gutenberg Editor Blocks test
    ✕ should be able to create a post with all blocks and scroll to the last one (80503 ms)

  ● Gutenberg Editor Blocks test › should be able to create a post with all blocks and scroll to the last one

    [performTouchAction([{"action":"press","options":{"x":720,"y":797.3333333333334}},{"action":"wait","options":{"ms":100}},{"action":"moveTo","options":{"x":720,"y":0}},{"action":"release","options":{}}])] Error response status: 12, , InvalidElementState - An element command could not be completed because the element is in an invalid state (e.g. attempting to click a disabled element). Selenium error: Swipe action cannot be performed

      at exports.newError (../../node_modules/wd/lib/utils.js:152:13)
      at ../../node_modules/wd/lib/callbacks.js:94:19
      at ../../node_modules/wd/lib/webdriver.js:202:5
      at Request._callback (../../node_modules/wd/lib/http-utils.js:89:7)
      at Request.self.callback (../../node_modules/request/request.js:185:22)
      at Request.<anonymous> (../../node_modules/request/request.js:1161:10)
      at IncomingMessage.<anonymous> (../../node_modules/request/request.js:1083:12)
```
Does it sound familiar? I'm not sure if I cherry-picked everything, maybe I missed something.

**Comment by geriux** [comment]
> @geriux So I cherry-picked all the commits from #43485 and it's been a great progress! The iOS e2e test suite is green, only the Android tests are failing. On this:

Thanks for cherry-picking the commits! I restarted the failed test as the E2E tests are a bit flaky at the moment and it passed correctly and all checks are green now 🚀

**Comment by jsnajdr** [comment]
> I restarted the failed test as the E2E tests are a bit flaky at the moment and it passed correctly and all checks are green now 🚀

Wow, does that mean we're mergeable now? 🙂

**Comment by geriux** [comment]
> Wow, does that mean we're mergeable now? 🙂

All good from the mobile side! :shipit:

**Review by geriux — APPROVED** [review]
All good from the mobile team 🚀

**Comment by tyxla** [comment]
As expected, I'm getting some of those:

> Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: [https://reactjs.org/link/switch-to-createroot](https://reactjs.org/link/switch-to-createroot)

but I've given this a round of testing in the various editors and everything appears to behave well so far. I think we should give this one a go!

**Comment by tyxla** [comment]
One thing we likely want to do beforehand is to add CHANGELOG entries where necessary. This is quite a notable change! Likely could use a [dev note](https://make.wordpress.org/core/handbook/tutorials/writing-developer-notes/) as well.

**Comment by jsnajdr** [comment]
> As expected, I'm getting some of those:
> > Warning: ReactDOM.render is no longer supported in React 18

Oh, this is something we should fix. But it will mean backward compatibility trouble for `@wordpress/element`. This package re-exports `render`, `hydrate` and `unmountComponentAtNode` from `react-dom`, but these functions are now deprecated. Replaced by `createRoot` and `hydrateRoot`.

What are we going to do? The current exports can be re-implemented using the React 18 API, keeping compatibility. Or breaking the compatibility and starting to export `createRoot` and `hydrateRoot` instead. What do you think @youknowriad?

> One thing we likely want to do beforehand is to add CHANGELOG entries where necessary.

What kind of entries do you have in mind? That the packages now depend on React 18? I think we didn't change anything else, like public API of anything.

**Comment by youknowriad** [comment]
> What are we going to do? The current exports can be re-implemented using the React 18 API, keeping compatibility. Or breaking the compatibility and starting to export createRoot and hydrateRoot instead. What do you think @youknowriad?

I think we should maintain backward compatibility as much as we can. We know for sure that upgrading to React 18 is not without issues so we should try to do it iteratively to avoid big disruptions. 

Also would be good to prepare a dev note with the potential breaking changes for the next WP release.

**Comment by Mamaduka** [comment]
Here's the PR that added notes for React 17 upgrade #33731.

**Comment by jsnajdr** [comment]
> Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17.

After thinking about this a bit more today, my conclusion is that it's a good thing we keep the React 17 mount APIs for now. That prevents the production Gutenberg from running in full concurrent mode, giving us some time to iron out concurrent bugs before really committing to it.

For now, only the unit tests are fully concurrent, because the mounting is done by `@testing-library/react` which already uses React 18 `createRoot` internally. The unit tests are one step ahead.

My biggest fear is that after merging this mega PR that changes 6000 lines in 120 files, we'll discover that the editor is broken and that we need to revert the upgrade. In the meantime other changes will introduce merge conflicts etc, a major headache.

**Comment by jsnajdr** [comment]
> One thing we likely want to do beforehand is to add CHANGELOG entries where necessary.

Added changelog entries to all packages where the React deps changes. I also found a few packages that declared a React peer dep despite not using React at all (like `notices` which contains just a data store).

I also tried to implement @noahtallen's suggestion to keep compatibility with React 17, i.e., declare React dependencies like `"^17.0 || ^18.0"`:
> In order for the rest of the ecosystem to update smoothly and at their own pace, should we support both 17 and 18 as peer deps in most packages? (As long as our code is still compatible with 17, that is)

But I gave up after a while. Some other downstream packages, like `framer-motion` used by `components`, already support only React 18. So, `components` cannot support React 17. Then any package that depends on `components` also can't realistically support React 17. Only a minor subset of all packages would really work with React 17.

It's better to do a big major release of everything that marks a clear point where React 17 support ends and React 18 begins.

**Comment by jsnajdr** [comment]
Hi @geriux :wave: I rebased this branch on top of #46211 which modified the `ios/Podfile.lock` file and I'm not sure if I resolved the conflicts right. Actually I'm sure I didn't because there are package hashes etc 🙂 Trying to run `bundle exec pod install` didn't work for me because I don't have the right Ruby gems installed and trying to install them gives me C++ compilation errors for Ruby extensions...

Could you please have a look at this branch and update the `Podfile.lock`?

**Comment by geriux** [comment]
> Could you please have a look at this branch and update the `Podfile.lock`?

Updated!

**Comment by youknowriad** [comment]
🎉 youhou

**Comment by tyxla** [comment]
Exciting to see this land 🚀

**Comment by Mamaduka** [comment]
Super excited to see this last. Thank you, @jsnajdr!

I assume the `Warning: ReactDOM.render is no longer supported in React 18` warning will be handled in follow-up PRs.

**Comment by tyxla** [comment]
> I assume the `Warning: ReactDOM.render is no longer supported in React 18` warning will be handled in follow-up PRs.

Yes. Technically, we're still using the non-concurrent mode and the old React 17 APIs that are still supported but yield a warning like this. We'll be updating to use the React 18 APIs incrementally.

**Comment by jsnajdr** [comment]
> Technically, we're still using the non-concurrent mode and the old React 17 APIs

Only the unit test suites are in full React 18 concurrent mode. In that sense, the tests are one step ahead 🙂 There is still some flakiness there and some remaining "not wrapped in `act()`" warnings. After ironing them out, we'll try to enable concurrent mode for the production editor.

**Comment by youknowriad** [comment]
There's a subtle change that was missed here:

In we remove this code https://github.com/WordPress/gutenberg/blob/a41ea8177b7ac9c2b1861bbacf5768ec459195b0/lib/compat/wordpress-6.0/client-assets.php#L19-L36  The react version that is on Core will be used instead. The problem is that code is in the `wordpress-6.0` folder (previous version we did a react upgrade) and this folder is going to be removed very soon because the minimum required version of WP in the Gutenberg plugin is going to be updated to 6.1 now. 

In other words, we need to move that code to `client-assets.php` of the 6.2 folder or to lib/client-assets (if we think we'll be upgrading the version often).

That said, we shouldn't forget to update the version in core as well, when updating the packages, so I'm adding the "backport to php" label as a reminder.

**Comment by danieliser** [comment]
So is the current version of all these packages that require React 18 not ready for consumption? Using any of them requires using all of them from my testing due to peerDependencies, unless you want to use the ignore flag every time which we don't.

Further pushing to actually try and use them resulted in typescript issues everywhere from underlying usage of @types/react with versions less than 18.

```
'CheckboxControl' cannot be used as a JSX component.
  Its element type 'ReactElement<any, any> | Component<Props, any, any> | null' is not a valid JSX element.
    Type 'Component<Props, any, any>' is not assignable to type 'Element | ElementClass | null'.
      Type 'Component<Props, any, any>' is not assignable to type 'ElementClass'.
        The types returned by 'render()' are incompatible between these types.
          Type 'React.ReactNode' is not assignable to type 'import("/home/daniel/contentcontrol/content-control/node_modules/@types/react/index").ReactNode'.
            Type '{}' is not assignable to type 'ReactNode'.ts(2786)
```

**Comment by jsnajdr** [comment]
> `Type 'React.ReactNode' is not assignable to type 'import("/home/.../node_modules/@types/react/index").ReactNode'`

This error indicates that there are two versions of `@types/react` being imported, the old and new one. Can you find where the old version is coming from? From your project's `package.json`, from some 3rd party library, from one of the `@types/wordpress__*` DefinitelyTyped packages?

> Using any of them requires using all of them

Unfortunately it was not feasible to keep compatibility with React 17. The `@wordpress/components` package, one of the fundamental ones, depends on 3rd party libraries that are already React 18 only, and that requirement would trickle down to all other packages. Discussed in this comment: https://github.com/WordPress/gutenberg/pull/45235#issuecomment-1331841906

**Comment by danieliser** [comment]
@jsnajdr I did track it down via `package-lock.json` to a few packages overall, I believe though that some of the `@types/wordpress__*` packages import `@wordpress/element` directly as part of its typings, due to this its currently locked on a version using React 17.

I know `@types/wordpress__compose` does this for sure, already started a patch, but haven't been able to test that it would actually work without further modifications.

https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...danieliser:DefinitelyTyped:master

**Comment by danieliser** [comment]
I believe it may be that outdated version of `@wordpress/element` being used in all the `@types/wordpress__*` packages that is causing the component type errors.

But this also mixes in with my other issue I created about the supposedly empty releases for all the packages that could in fact contain breaking changes that are not getting documented:  #46730

**Comment by danieliser** [comment]
Appears to be coming from `@types/wordpress__components`

image

Will test overrides locally to see if simply changing its deps to use the newer `@wordpress/element@5` package fixes it.

**Comment by danieliser** [comment]
Confirmed, simply changing that dependency in the `@types/wordpress__components` package fixed the bulk of the TS errors.

There are a few straggling errors that didn't exist before but there are only 6 and I'll figure those out shortly.

**Comment by danieliser** [comment]
Ugg, just me or is that types package woefully untested.

Many of the type files in it import packages not in package.json.

Am I missing something, or should those all be listed?

Had to add the following to get it to pass `npm test wordpress__components`

```
"@types/wordpress__rich-text": "^3.4.6",
"@types/tinycolor2": "^1.4.3",
"@types/wordpress__notices": "3.5.0",
```

**Comment by danieliser** [comment]
Also, what branch are the current releases in? `trunk` does not contain the "current" npm release for @wordpress/components @23.0.0.

Trunk shows those changes which are clearly live for the world as `Unreleased`

**Comment by danieliser** [comment]
PR submitted that seemed to resolve it for me, no idea if/when it gets approved, though would love some extra testers: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63716

**Comment by Mamaduka** [comment]
I've removed backport label. The changes will be part of the general package update.

**Comment by gziolo** [comment]
@desrosj discovered some issues with peer dependencies defined in some React libraries when testing WordPress core with Node 18.14.0 and npm: 9.3.1: https://github.com/WordPress/wordpress-develop/pull/4028. I filed an issue  https://github.com/WordPress/gutenberg/issues/48009 in Gutenberg with steps to reproduce.
